File: frontend/client/src/types/architecture.ts
export type ArchitectureStatus =
  | "READY_FOR_REVIEW"
  | "DRAFT_REVISION"
  | "APPROVED"
  | "NEEDS_CLARIFICATION"
  | "UNSUPPORTED";
export type Environment = "development" | "staging" | "production";
export type BudgetPreference = "MINIMIZE_COST" | "BALANCED" | "PERFORMANCE_FIRST";
export type AvailabilityRequirement = "STANDARD" | "HIGH" | "CRITICAL";
export type CloudProvider = "AWS";

export interface RequirementContext {
  environment: Environment;
  budget_preference: string;
  cloud: CloudProvider;
  region?: string;
}

export interface ArchitectureComponent {
  id: string;
  name: string;
  type?: string;
  provider_type?: string;
  category: string;
  purpose: string;
  scope: string;
  configuration?: Record<string, unknown>;
  dependencies?: string[];
  depends_on?: string[];
  diagram_visibility?: string;
}

export interface ArchitectureDecision {
  id?: string;
  title: string;
  decision?: string;
  rationale: string;
  alternatives?: Array<{
    name: string;
    pros: string[];
    cons: string[];
    reason_not_selected: string;
  }>;
  alternatives_considered?: Array<{
    option: string;
    advantages: string[];
    disadvantages: string[];
    reason_not_selected: string;
  }>;
}

export interface ArchitectureScore {
  security: string;
  costEfficiency: string;
  operationalComplexity: string;
  scalability: string;
}

export interface CanonicalArchitecture {
  schema_version: string;
  architecture_id: string;
  architecture_version: number | string;
  status: ArchitectureStatus;
  title: string;
  requirement_summary: {
    business_goal: string;
    application_type: string;
    environment: Environment;
    expected_users_or_traffic: string;
    functional_requirements: string[];
    non_functional_requirements: string[];
    constraints: string[];
    budget_preference?: string;
    availability_requirement?: string;
  };
  cloud: {
    provider: CloudProvider;
    region: string;
    region_rationale: string;
  };
  solution: string | {
    summary: string;
    architecture_pattern: string;
  };
  resources: ArchitectureComponent[];
  relationships: Array<{
    source?: string;
    target?: string;
    type?: string;
    source_id?: string;
    target_id?: string;
    relationship_type?: string;
    label?: string;
  }>;
  decisions: ArchitectureDecision[];
  security_considerations: string[];
  reliability_considerations: string[];
  scalability_considerations: string[];
  cost_considerations: string[];
  operational_considerations: string[];
  assumptions: string[];
  open_questions: string[];
  risks: Array<{
    risk?: string;
    title?: string;
    description?: string;
    severity?: string;
    mitigation: string;
  }>;
  limitations: string[];
  recommended_diagrams: string[];
  markdown_report?: string;
  high_level_diagram?: string;
  network_diagram?: string;
}

export interface Project {
  id: string;
  title: string;
  requirement: string;
  context: RequirementContext;
  architecture: CanonicalArchitecture;
  status: ArchitectureStatus;
  createdAt: Date;
  updatedAt: Date;
  resourceCount: number;
}

export interface AnalyzeArchitectureRequest {
  requirement: string;
  context: RequirementContext;
}

export interface AnalyzeArchitectureResponse {
  architecture: CanonicalArchitecture;
  report_markdown: string;
  metadata: {
    provider: string;
    model: string;
    generation_duration_ms: number;
  };
}

export interface BackendProject {
  id: string;
  title: string;
  slug: string;
  initialRequirement: string;
  context: Record<string, unknown>;
  currentVersionId: string | null;
  status: ArchitectureStatus;
  userId: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface BackendArchitectureVersion {
  id: string;
  projectId: string;
  version: string;
  parentVersionId: string | null;
  status: ArchitectureStatus;
  architecture: CanonicalArchitecture;
  reportMarkdown: string;
  simpleDiagramMermaid: string | null;
  advancedDiagramMermaid: string | null;
  changeSummary: string[];
  metadata: Record<string, unknown>;
  createdAt: string;
}

export interface BackendChatMessage {
  id: string;
  projectId: string;
  architectureVersionId: string | null;
  role: "user" | "assistant" | "system";
  content: string;
  intent: "INITIAL" | "EXPLAIN" | "MODIFY" | "CLARIFY" | "UNSUPPORTED" | "SYSTEM";
  architectureChanged: boolean;
  createdAt: string;
}

export interface BackendChangeRequest {
  id: string;
  projectId: string;
  fromVersionId: string;
  toVersionId: string | null;
  userMessage: string;
  changeSummary: string[];
  status: "pending" | "accepted" | "discarded";
  createdAt: string;
  updatedAt: string;
}

export interface ProjectWorkspaceResponse {
  project: BackendProject;
  currentVersion: BackendArchitectureVersion | null;
  versions: BackendArchitectureVersion[];
  messages: BackendChatMessage[];
}

export interface ProjectMessageResponse {
  intent: "EXPLAIN" | "MODIFY" | "CLARIFY" | "UNSUPPORTED";
  architectureChanged: boolean;
  answer: string;
  messages: BackendChatMessage[];
  project?: BackendProject | null;
  currentVersion?: BackendArchitectureVersion | null;
  draftVersion?: BackendArchitectureVersion | null;
  changeRequest?: BackendChangeRequest | null;
  changeSummary: string[];
}

File: agents/cloud-architect-agent/app/schemas/architecture.py
from enum import Enum
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field, field_validator, model_validator

class ArchitectureStatus(str, Enum):
    NEEDS_CLARIFICATION = "NEEDS_CLARIFICATION"
    READY_FOR_REVIEW = "READY_FOR_REVIEW"
    UNSUPPORTED = "UNSUPPORTED"

class ResourceCategory(str, Enum):
    NETWORK = "NETWORK"
    COMPUTE = "COMPUTE"
    CONTAINER = "CONTAINER"
    DATABASE = "DATABASE"
    STORAGE = "STORAGE"
    SECURITY = "SECURITY"
    LOAD_BALANCING = "LOAD_BALANCING"
    OBSERVABILITY = "OBSERVABILITY"
    IDENTITY = "IDENTITY"
    OTHER = "OTHER"

class ResourceScope(str, Enum):
    GLOBAL = "GLOBAL"
    REGIONAL = "REGIONAL"
    VPC = "VPC"
    PUBLIC_SUBNET = "PUBLIC_SUBNET"
    PRIVATE_SUBNET = "PRIVATE_SUBNET"
    PUBLIC = "PUBLIC"
    PRIVATE = "PRIVATE"

class RelationshipType(str, Enum):
    ROUTES_TRAFFIC_TO = "ROUTES_TRAFFIC_TO"
    CONNECTS_TO = "CONNECTS_TO"
    READS_FROM = "READS_FROM"
    WRITES_TO = "WRITES_TO"
    AUTHENTICATES_WITH = "AUTHENTICATES_WITH"
    MONITORS = "MONITORS"
    DEPLOYS_TO = "DEPLOYS_TO"
    DEPENDS_ON = "DEPENDS_ON"

class RiskSeverity(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"

class Alternative(BaseModel):
    option: str
    advantages: List[str]
    disadvantages: List[str]
    reason_not_selected: str

class Decision(BaseModel):
    title: str
    decision: str
    rationale: str
    alternatives_considered: List[Alternative]

class Risk(BaseModel):
    title: str
    description: str
    severity: RiskSeverity
    mitigation: str

class Resource(BaseModel):
    id: str = Field(..., pattern=r"^[a-z0-9_-]+$")
    name: str
    provider_type: str
    category: ResourceCategory
    scope: ResourceScope
    purpose: str
    configuration: Dict[str, Any]
    depends_on: List[str] = []

class Relationship(BaseModel):
    source_id: str
    target_id: str
    relationship_type: RelationshipType
    label: str

class RequirementSummary(BaseModel):
    business_goal: str
    application_type: str
    environment: str
    expected_users_or_traffic: Optional[str] = None
    functional_requirements: List[str]
    non_functional_requirements: List[str]
    constraints: List[str]
    budget_preference: Optional[str] = None
    availability_requirement: Optional[str] = None

class CloudConfig(BaseModel):
    provider: str = "AWS"
    region: str
    region_rationale: str

class ArchitectureSpecification(BaseModel):
    schema_version: str = "1.0"
    architecture_id: str
    architecture_version: str = "1.0"
    status: ArchitectureStatus
    title: str
    requirement_summary: RequirementSummary
    cloud: CloudConfig
    solution: str
    resources: List[Resource] = []
    relationships: List[Relationship] = []
    decisions: List[Decision] = []
    security_considerations: List[str] = []
    reliability_considerations: List[str] = []
    scalability_considerations: List[str] = []
    cost_considerations: List[str] = []
    operational_considerations: List[str] = []
    assumptions: List[str] = []
    open_questions: List[str] = []
    risks: List[Risk] = []
    limitations: List[str] = []
    recommended_diagrams: List[str] = ["high-level"]

    @model_validator(mode="after")
    def validate_logic(self) -> "ArchitectureSpecification":
        # Check resource IDs uniqueness
        resource_ids = [r.id for r in self.resources]
        if len(resource_ids) != len(set(resource_ids)):
            raise ValueError("Resource IDs must be unique")
            
        # Check relationships and dependencies
        for rel in self.relationships:
            if rel.source_id not in resource_ids:
                raise ValueError(f"Relationship source '{rel.source_id}' does not exist")
            if rel.target_id not in resource_ids:
                raise ValueError(f"Relationship target '{rel.target_id}' does not exist")
                
        for res in self.resources:
            for dep in res.depends_on:
                if dep not in resource_ids:
                    raise ValueError(f"Dependency '{dep}' for resource '{res.id}' does not exist")
                    
        # Status specific checks
        if self.status == ArchitectureStatus.READY_FOR_REVIEW and not self.resources:
            raise ValueError("READY_FOR_REVIEW status must contain at least one resource")
        if self.status == ArchitectureStatus.NEEDS_CLARIFICATION and not self.open_questions:
            raise ValueError("NEEDS_CLARIFICATION status must contain at least one open question")
        if self.status == ArchitectureStatus.UNSUPPORTED and not self.limitations:
            raise ValueError("UNSUPPORTED status must contain an explanation in limitations or solution")
            
        return self
